neuralmonkey/vocabulary.py

```diff
@@ -379,8 +379,13 @@ def truncate(self, size: int) -> None:
                                key=lambda w: self.word_count[w])
 
         # keep the least frequent words which are not special symbols
-        words_to_delete = [w for w in words_by_freq[:-size]
-                           if not _is_special_token(w)]
+        deleted_size = self.__len__() - size
+        words_to_delete = []
+        for word in words_by_freq:
+            if not _is_special_token(word):
+                words_to_delete.append(word)
+            if len(words_to_delete) == deleted_size:
+                break
```

Comment:
je to to chování co chceme? zvýhodňuje to slova začínající na písmenka ke konci abecedy.

Nechcem vyhodit všechna slova s danou frekvencí? (otázka i na @jlibovicky)

Reply:
Pokud chces vyhazovat slova pod urcitou hranici frequence, tak pouzivej tohle
https://github.com/ufal/neuralmonkey/blob/master/neuralmonkey/vocabulary.py#L399

Reply:
To já vim že je na to funkce, ale tu jednak nemůžeš použít, když nevíš, jakej bude teda ten threshold, aby se ti do slovníku vešlo tolik a tolik slov, a jednak jak jsem řekl, rozhodování, co do toho slovníku spadne a co ne je dost arbitrární a ptám se, jestli by se nemělo dělat nějak chytřejš.

Reply:
No ted kdyz to je podle abecedy, tak je to aspon deterministicke.

Reply:
To jsou fakta to co píšeš já se ptal na názory

Reply:
No muj nazor je, ze proste to truncatujes na pevnou size, takze holt na te hranici nektera slova vyhodis a nektera nevyhodis. Zivot je kruty, se s tim smir :)
Podstatnejsi podle me je, ze ta funkce dodrzi vracenou vocabulary_size a ze se na stejnych datech/parametrech bude vzdy chovat stejne.

Reply:
Ale k čemu takovej experiment je? Vědec na to kouká takhle: "chci tady slovník a chci aby v něm bylo maximálně 50k slov a chci vyhodit vzácný slova." To, že tam bude "zilión" a "zebřišti", a nebude tam "andělíčků" a "bambilión", ačkoliv maj stejnou frekvenci (a tim pádem i pravděpodobnost výskytu), to ho ani nenapadne a ani to asi tak nechce. A model tim pádem bude dávat špatný odhad pravděpodobnosti větě "Zilión zeber na zebřišti" v porovnání třeba s větou "Bambilión zeber na zebřišti".

Reply:
Btw. jinak tohle by bylo taky deterministický..

Reply:
A vocabulary_size to nedodržuje protože se tam šaškuje se special symbolama

Reply:
Vocabulary size to dodrzuje, protoze mas predem odhadnuto, kolik znaku chces smaznout aby ti zustal slovnik o velikosti 'size' (vcetne specia_tokens).

Reply:
Tak klidne muzeme vyhazet vsechny slova od dane frekvence, ale pak bych tam minimalne pridal warning (se skutecnou finalni velikosti slovniku), aby o tom user vedel.
Btw, jak chces resit situaci, kdy chces truncatnout napr. na size=500 a zrovna distribuce frekvenci je takova, ze ti vyjde budto slovnik velikosti ~100 nebo ~900. Zrejme nechces ani jeden pripad.

Reply:
to je ale imho dost nereálnej příklad se kterym bys nic rozumnýho nenatrénoval

Reply:
tady se bavíme o tom, jestli má mít slovník přesně 50000 slov nebo 46521, případně 58021

Reply:
(i tohle je asi dost nereálná situace ale nevim, jak ten long tail může vypadat)